src/v/datalake/coordinator/coordinator.cc

```diff
@@ -47,11 +54,147 @@ std::ostream& operator<<(std::ostream& o, coordinator::errc e) {
     }
 }
 
+void coordinator::start() {
+    ssx::spawn_with_gate(gate_, [this] { return run_until_abort(); });
+}
+
 ss::future<> coordinator::stop_and_wait() {
     as_.request_abort();
+    leader_cond_.broken();
     return gate_.close();
 }
 
+ss::future<> coordinator::run_until_abort() {
+    while (!as_.abort_requested()) {
+        auto& raft = *stm_->raft();
+        if (!raft.is_leader()) {
+            bool shutdown = false;
+            auto leader = co_await ss::coroutine::as_future(
```

Comment:
I think some of the complexity in these nested loops comes from exception handling, elsewhere I had a similar problem and I ended up using the following pattern.

ssx::repeatuntil_gate_closed_or_aborted( _gate, _as,  {do_action_once()}.handle_exception() { log_failure();}) 

Just throw away all state and retry again until the gate/as is active.